### PR TITLE
Add better error handling for PPOPRF

### PR DESCRIPTION
**Pull request by DJAndries**
@rillian I would assume we would want to make a separate upstream PR for these changes. If you agree, then could you create a `better-error-handling` branch? I can change this PR to merge into that branch instead.

### CODE REVIEW

File: ppoprf/src/ggm.rs [hunk 1]
```diff
@@ -5,19 +5,13 @@
...
-use super::PPRF;
+use super::{PPRF, PPRFError};
 use bitvec::prelude::*;
 use ring::{
     hmac,
     rand::{SecureRandom, SystemRandom},
 };
 
-#[derive(Debug)]
-enum GGMError {
```
Comment (rillian) [inline ppoprf/src/ggm.rs:16]: It feels weird for the `ggm` module to use the upstream trait's error type outside of implementing that type. Is there too much duplication if you keep `GGMError` and just impl `From` or `Into` for it?

File: ppoprf/src/lib.rs [hunk 2]
```diff
@@ -15,8 +15,22 @@
...
+#[derive(Debug, Error, Display, PartialEq)]
+pub enum PPRFError {
+    #[display(fmt = "Tag index is out of bounds, indicated index {} is above length of {}", index, tag_size)]
+    BadTagIndex { index: usize, tag_size: usize },
+    #[display(fmt = "No prefix found")]
+    NoPrefixFound,
+    #[display(fmt = "Tag already punctured")]
+    AlreadyPunctured,
+    #[display(fmt = "Input length ({}) does not match input param ({})", actual, expected)]
+    BadInputLength { actual: usize, expected: usize }
```
Comment (rillian) [inline ppoprf/src/lib.rs:29]: `BadInputValue` may be possible too, if the point is off-curve? I suppose that can be added later.

**Review by rillian — COMMENTED** [review]
I'm all for returning results instead of panic! on bad arguments, at least with the public functions! Since `eval` and `puncture` take raw buffers for crypto values there's no way to make them infallible in the trait.

This isn't really my code though, so I'd suggest submitting the changes to the ppoprf crate as a separate upstream PR, like you mentioned. Once we've gotten feedback on that we can worry about propagating the new errors in the example server.

**Comment by DJAndries** [comment]
Closing, will create upstream PR once @rillian gets his server example changes merged.

**Pull request closed without merging**